stats/statsvariables.cpp

```diff
@@ -0,0 +1,1674 @@
+template<typename Binner, typename Bin>
+struct MultiBinner : public StatsBinner {
+public:
+	using Type = decltype(Bin::value);
+	std::vector<StatsBinDives> bin_dives(const std::vector<dive *> &dives, bool fill_empty) const override;
+	std::vector<StatsBinCount> count_dives(const std::vector<dive *> &dives, bool fill_empty) const override;
+	const Binner &derived() const {
+		return static_cast<const Binner &>(*this);
+	}
+	const Bin &derived_bin(const StatsBin &bin) const {
+		return dynamic_cast<const Bin &>(bin);
+	}
+};
+
+template<typename Binner, typename Bin>
+std::vector<StatsBinDives> MultiBinner<Binner, Bin>::bin_dives(const std::vector<dive *> &dives, bool) const
+{
+	// First, collect a value / dives vector and then produce the final vector
+	// out of that. I wonder if that is permature optimization?
+	using Pair = std::pair<Type, std::vector<dive *>>;
+	std::vector<Pair> value_bins;
+	for (dive *d: dives) {
+		for (const Type &val: derived().to_bin_values(d)) {
+			if (is_invalid_value(val))
+				continue;
+			register_bin_value(value_bins, val,
+					   [d](std::vector<dive *> &v) { v.push_back(d); });
+		}
+	}
+
+	// Now, turn that into our result array with allocated bin objects.
+	return value_vector_to_bin_vector<Bin>(*this, value_bins, false);
+}
+
+template<typename Binner, typename Bin>
+std::vector<StatsBinCount> MultiBinner<Binner, Bin>::count_dives(const std::vector<dive *> &dives, bool) const
+{
+	// First, collect a value / counts vector and then produce the final vector
+	// out of that. I wonder if that is permature optimization?
+	using Pair = std::pair<Type, int>;
+	std::vector<Pair> value_bins;
+	for (const dive *d: dives) {
+		for (const Type &s: derived().to_bin_values(d)) {
+			if (is_invalid_value(s))
+				continue;
+			register_bin_value(value_bins, s, [](int &i){ ++i; });
+		}
+	}
+
+	// Now, turn that into our result array with allocated bin objects.
+	return value_vector_to_bin_vector<Bin>(*this, value_bins, false);
+}
+
+// A binner that works on string-based bins whereby each dive can
+// produce multiple strings (e.g. dive buddies). The binner must
+// feature a to_bin_values() function that produces a vector of
+// QStrings and bins that can be constructed from QStrings.
+// Other than that, see SimpleBinner.
+template<typename Binner, typename Bin>
+struct StringBinner : public MultiBinner<Binner, Bin> {
+public:
+	QString format(const StatsBin &bin) const override {
+		return dynamic_cast<const Bin &>(bin).value;
+	}
+};
+
+// ============ The date of the dive by year, quarter or month ============
+// (Note that calendar week is defined differently in different parts of the world and therefore omitted for now)
+
+double date_to_double(int year, int month, int day)
+{
+	struct tm tm = { 0 };
+	tm.tm_year = year;
+	tm.tm_mon = month;
+	tm.tm_mday = day;
+	timestamp_t t = utc_mktime(&tm);
+	return t / 86400.0; // Turn seconds since 1970 to days since 1970, if that makes sense...?
```

Comment:
stupid bike shedding comment... leap seconds...?

Reply:
I'm not qualified enough to make a comment on that. I.e. I don't know how leap seconds are treated by the C library.

In any case, the deviation from linearity should be negligible.

Reply:
It was meant as a joke among all the others... mainly because you divided by `86400.0` which of course was intended to force floating point math, but to my silly eyes looked like a statement of precision...